drivers/usb/udc/udc_ambiq.c

```diff
@@ -0,0 +1,915 @@
+	}
+
+	udc_ep_set_busy(dev, ep_cfg->addr, false);
+	am_hal_usb_ep_state_reset(priv->usbHandle, ep_cfg->addr);
+	irq_unlock(lock_key);
+
+	LOG_DBG("dequeue ep 0x%02x", ep_cfg->addr);
+
+	return 0;
+}
+
+/* Halt endpoint */
+static int udc_ambiq_ep_set_halt(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("Halt ep 0x%02x", ep_cfg->addr);
+
+	am_hal_usb_ep_stall(priv->usbHandle, ep_cfg->addr);
+
+	return 0;
+}
+
+/* Clear EP halt */
+static int udc_ambiq_ep_clear_halt(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("Clear halt ep 0x%02x", ep_cfg->addr);
+
+	am_hal_usb_ep_clear_stall(priv->usbHandle, ep_cfg->addr);
+
+	return 0;
+}
+
+/* Configure and enable endpoint */
+static int udc_ambiq_ep_enable(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+	uint8_t endpoint_type;
+	uint32_t status;
+
+	__ASSERT_NO_MSG(ep_cfg);
+
+	switch (ep_cfg->attributes) {
+	case USB_EP_TYPE_CONTROL:
+		endpoint_type = 0; /* AM_HAL_USB_EP_XFER_CONTROL */
+		break;
+	case USB_EP_TYPE_ISO:
+		endpoint_type = 1; /* AM_HAL_USB_EP_XFER_ISOCHRONOUS */
+		break;
+	case USB_EP_TYPE_BULK:
+		endpoint_type = 2; /* AM_HAL_USB_EP_XFER_BULK */
+		break;
+	case USB_EP_TYPE_INTERRUPT:
+		endpoint_type = 3; /* AM_HAL_USB_EP_XFER_INTERRUPT */
+		break;
+	default:
+		return -EINVAL;
+	}
+	status = am_hal_usb_ep_init(priv->usbHandle, ep_cfg->addr, endpoint_type, ep_cfg->mps);
+	if (status != AM_HAL_STATUS_SUCCESS) {
+		LOG_ERR("am_hal_usb_ep_init failed(0x%02x), %d", ep_cfg->addr, (int)status);
+		return -EIO;
+	}
+	LOG_DBG("Enable ep 0x%02x", ep_cfg->addr);
+
+	return 0;
+}
+
+/* Disable endpoint */
+static int udc_ambiq_ep_disable(const struct device *dev, struct udc_ep_config *ep_cfg)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	__ASSERT_NO_MSG(ep_cfg);
+	am_hal_usb_ep_stall(priv->usbHandle, ep_cfg->addr);
+	LOG_DBG("Disable ep 0x%02x", ep_cfg->addr);
+
+	return 0;
+}
+
+/* Initiate host wakeup procedure */
+static int udc_ambiq_host_wakeup(const struct device *dev)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	am_hal_usb_start_remote_wakeup(priv->usbHandle);
+
+	return 0;
+}
+
+/* Set USB device address */
+static int udc_ambiq_set_address(const struct device *dev, const uint8_t addr)
+{
+	struct udc_ambiq_data *priv = udc_get_private(dev);
+
+	LOG_DBG("addr %u (0x%02x)", addr, addr);
+	am_hal_usb_set_addr(priv->usbHandle, addr);
+	am_hal_usb_set_dev_state(priv->usbHandle, AM_HAL_USB_DEV_STATE_ADDRESSED);
+
+	return 0;
+}
+
+/* Enable Test Mode */
+static int udc_ambiq_test_mode(const struct device *dev, const uint8_t mode, const bool dryrun)
+{
+	LOG_DBG("Enter test mode.");
+
+	// TODO: Study and add TestMode support or remove test mode from interface
+
+	__ASSERT(0, "Test mode not supported");
```

Comment:
Why is there ASSERT? ` * @retval -ENOTSUP Test mode is not supported` !

Reply:
Hi @jfischer-no, sorry for missing out on the porting of test mode earlier. 
The actual test mode code is now uploaded. This should no longer be an issue. 😉